### PR TITLE
ddl: fix concurrent column type changes(with changing data) that cause schema and data inconsistencies

### LINKED ISSUE #31048: Concurrent Column Type Change would cause inconsistent schema and data
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

``` SQL
create table test(a int, b int);
insert into test values (1, 1);
alter table test change column b b double; /* in session 1 */
alter table test change column b b double; /* in session 2 */
select * from test;
```

### 2. What did you expect to see? (Required)
No error

### 3. What did you see instead (Required)
``` SQL
mysql> select * from test;
ERROR 1105 (HY000): insufficient bytes to decode value
```
### 4. What is your TiDB version? (Required)
5.1 5.2 5.3 master

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve?

Issue Number: close #31048

Problem Summary:

For column type changes, the column ID is changed if it is lossy.

If both sessions perform lossy column type changes((like modify column c int -> char)) simultaneously

1. Connection1: execute DDL1, it will change column ID (ID :1-> ID :2)

2. Connection2: execute DDL2. After DDL1 enters the DDL job queue and is not executed, then DDL2 enters the job queue.

3. After DDL1 is executed successfully, DDL2 considers that the column is not lossy(char -> char), so it does not change the column ID, which is the old id:1. So id:1 overwrites id:2.

After DDL2 is completed, the column ID is still the original id:1 instead of the correct id:2.

Because the column ID was changed incorrectly, the schema could not match the data, resulting in decode failure when reading data.

### What is changed and how it works?
In step 3, we check that the ID of the modified column is the same as the ID of the original column.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fix concurrent column type changes(with changing data) that cause schema and data inconsistencies.
```

### CODE REVIEW

**Review by AilinKid — APPROVED** [review]
LGTM

File: ddl/db_change_test.go [hunk 1]
```diff
@@ -1062,6 +1062,54 @@ func (s *testStateChangeSuite) TestParallelAlterModifyColumn(c *C) {
 	s.testControlParallelExecSQL(c, sql, sql, f)
 }
 
+func (s *testStateChangeSuite) TestParallelAlterModifyColumnWithData(c *C) {
```
Comment (wjhuang2016) [inline ddl/db_change_test.go:1065]: Can we add more tests about `concurrent alter table`? For example, random test.

Comment (zimulala) [reply ddl/db_change_test.go:1065]: Could you give an example of this random test? I don't think it's appropriate to add it here.

**Review by tangenta — APPROVED** [review]

File: ddl/column.go [hunk 2]
```diff
@@ -1483,6 +1483,10 @@ func updateChangingInfo(changingCol *model.ColumnInfo, changingIdxs []*model.Ind
 func (w *worker) doModifyColumn(
 	d *ddlCtx, t *meta.Meta, job *model.Job, dbInfo *model.DBInfo, tblInfo *model.TableInfo,
 	newCol, oldCol *model.ColumnInfo, pos *ast.ColumnPosition) (ver int64, _ error) {
+	if oldCol.ID != newCol.ID {
+		job.State = model.JobStateRollingback
+		return ver, errKeyColumnDoesNotExits.GenWithStack("column %s id %d does not exist, column id may be updated by other DDL ran in parallel", oldCol.Name, newCol.ID)
```
Comment (bb7133) [inline ddl/column.go:1488]:
```suggestion
		return ver, errKeyColumnDoesNotExits.GenWithStack("Column %s with ID %d does not exist, this column may have been updated by other DDL ran in parallel", oldCol.Name, newCol.ID)
```
How about this?

**Comment by zimulala** [comment]
PTAL @wjhuang2016 @bb7133

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**